src/s390x/zipl.rs

```diff
@@ -29,22 +40,196 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
     Ok(())
 }
 
-/// Runs `zipl` based on Ignition and BLS configuration in `boot`.
-pub fn zipl<P: AsRef<Path>>(boot: P) -> Result<()> {
-    let boot = boot.as_ref();
+fn is_secure_execution() -> bool {
+    if let Ok(s) = std::fs::read_to_string("/sys/firmware/uv/prot_virt_guest") {
+        if s.trim() == "1" {
+            return true;
+        }
+    }
+    false
+}
 
-    // create dummy config for zipl
-    let mut conffile = Builder::new()
-        .prefix("coreos-installer-zipl.")
+fn find_files<P: AsRef<Path> + ?Sized>(
+    config_path: &P,
+    f: impl Fn(&PathBuf) -> bool,
+) -> Result<Vec<PathBuf>> {
+    let mut entries = read_dir(&config_path)
+        .with_context(|| format!("reading directory {}", config_path.as_ref().display()))?
+        .filter_map(Result::ok)
+        .filter_map(|e| if f(&e.path()) { Some(e.path()) } else { None })
+        .collect::<Vec<_>>();
+    entries.sort();
+    Ok(entries)
+}
+
+fn get_bls_entry_variable<P: AsRef<Path>>(boot: P, name: &str) -> Result<String> {
+    let bls_path = boot.as_ref().join("loader/entries");
+    let mut entries = find_files(&bls_path, |p: &PathBuf| {
+        p.extension().unwrap_or_default() == "conf"
+    })?;
+    let path = entries
+        .pop()
+        .with_context(|| format!("Found no BLS entries in {}", bls_path.display()))?;
+    let contents =
+        std::fs::read_to_string(&path).with_context(|| format!("reading {}", path.display()))?;
+    let name = format!("{} ", name.trim());
+    let line = contents
+        .lines()
+        .find(|l| l.starts_with(&name))
+        .with_context(|| format!("'{}' doesn't have '{}'", path.display(), name))?;
+    Ok(line[name.len()..].trim().to_owned())
+}
+
+fn regenerate_initrd<P: AsRef<Path>>(path: P, files: &[PathBuf]) -> Result<NamedTempFile> {
```

Comment:
Similarly to the cosa PR, I think here rather than decompressing the initrd and appending, we can just append the little bit we need to a copy of the initrd.

Does the initrd in `/boot` need to have the LUKS keys? IIUC, it only actually needs to be in the initrd that becomes part of `sdboot`, right? So we could just leave the canonical initrd alone. That way, we don't need to think about possibly appending the same thing over and over.

Reply:
> i've tried to do so. but that didn't work.

OK, let's debug that bit in the cosa PR first and then circle back here.

> what do you mean? `rdcore` is called just once on firstboot-complete, so luks keys are appended just once

Yes, you're right. Though in the spirit of `rdcore` having re-usable pieces, I think we should leave the possibility open for `rdcore zipl` being called multiple times. Anyway, it sounds like this won't be an issue since we don't have to modify the initrd in `/boot`.

Reply:
done. looks like my issue was lack of `-mindepth 1` and `/etc/luks` in `cpio`